timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java

```diff
@@ -98,34 +98,38 @@ private static void advanceAtomicLongToValue(AtomicLong toAdvance, long val) {
-                        if (allocationFailure instanceof MultipleRunningTimestampServiceError) {
-                            // We cannot allocate timestamps anymore because another server is running.
-                            return;
-                        }
-                        allocateMoreTimestamps();
-                        lastAllocatedTime = clock.getTimeMillis();
-                        allocationFailure = null;
-                    } catch (Throwable e) { // (authorized)
-                        createdException.initCause(e);
-                        if (allocationFailure != null
-                                && e.getClass().equals(allocationFailure.getClass())) {
-                            // QA-75825: don't keep logging error if we keep failing to allocate.
-                            log.info("Throwable while allocating timestamps.", createdException);
-                        } else {
-                            log.error("Throwable while allocating timestamps.", createdException);
+        if (isAllocationTaskSubmitted.compareAndSet(false, true)) {
+            if (isAllocationRequired(lastReturnedTimestamp.get(), upperLimitToHandOutInclusive.get())) {
+                final Exception createdException = new Exception("allocation task called from here");
+                executor.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        try {
+                            if (allocationFailure instanceof MultipleRunningTimestampServiceError) {
+                                // We cannot allocate timestamps anymore because another server is running.
+                                return;
+                            }
+                            allocateMoreTimestamps();
+                            lastAllocatedTime = clock.getTimeMillis();
+                            allocationFailure = null;
+                        } catch (Throwable e) { // (authorized)
+                            createdException.initCause(e);
+                            if (allocationFailure != null
+                                    && e.getClass().equals(allocationFailure.getClass())) {
+                                // QA-75825: don't keep logging error if we keep failing to allocate.
+                                log.info("Throwable while allocating timestamps.", createdException);
+                            } else {
+                                log.error("Throwable while allocating timestamps.", createdException);
+                            }
+                            allocationFailure = e;
+                        } finally {
+                            isAllocationTaskSubmitted.set(false);
                         }
-                        allocationFailure = e;
-                    } finally {
-                        isAllocationTaskSubmitted.set(false);
                     }
-                }
-            });
+                });
```

Comment:
That was my first thought too (see earlier commit on this branch).  Sadly, reordering the check makes it possible for this to run multiple times, which is not a disaster in practice but can lead to test flakiness. 

The clean/proper solution will be to let java handle the concurrency aspect for us, per #466 (which is already in progress).  Since that's coming soon anyway, I'm happy with this patched solution.